liquid-compiler/src/parser.rs

```diff
@@ -3,622 +3,406 @@
 //! This module contains functions than can be used for writing plugins
 //! but should be ignored for simple usage.
 
-use std::collections::HashSet;
-use std::iter::FromIterator;
-use std::slice::Iter;
-
 use liquid_interpreter::Expression;
 use liquid_interpreter::Renderable;
 use liquid_interpreter::Text;
 use liquid_interpreter::Variable;
 use liquid_interpreter::{FilterCall, FilterChain};
+use liquid_value::Scalar;
+use liquid_value::Value;
 
 use super::error::{Error, Result};
-use super::Element;
 use super::LiquidOptions;
 use super::ParseBlock;
 use super::ParseTag;
-use super::Token;
-
-/// Parses the provided elements into a number of Renderable items
-/// This is the internal version of parse that accepts Elements tokenized
-/// by `lexer::tokenize` and does not register built-in blocks. The main use
-/// for this function is for writing custom blocks.
-///
-/// For parsing from a String you should refer to `liquid::parse`.
-pub fn parse(elements: &[Element], options: &LiquidOptions) -> Result<Vec<Box<Renderable>>> {
-    let mut ret = vec![];
-    let mut iter = elements.iter();
-    let mut token = iter.next();
-    while token.is_some() {
-        let render = match *token.unwrap() {
-            Element::Expression(ref tokens, _) => parse_expression(tokens, options)?,
-            Element::Tag(ref tokens, _) => parse_tag(&mut iter, tokens, options)?,
-            Element::Raw(ref x) => Box::new(Text::new(x.as_str())),
-        };
-        ret.push(render);
-        token = iter.next();
-    }
-    Ok(ret)
-}
 
-const NOTHING: Option<&str> = None;
+use pest::Parser;
 
-pub fn unexpected_token_error<S: ToString>(expected: &str, actual: Option<S>) -> Error {
-    let actual = actual.map(|x| x.to_string());
-    unexpected_token_error_string(expected, actual)
-}
+#[derive(Parser)]
+#[grammar = "grammar.pest"]
+struct LiquidParser;
 
-pub fn unexpected_token_error_string(expected: &str, actual: Option<String>) -> Error {
-    let actual = actual.unwrap_or_else(|| "nothing".to_owned());
-    Error::with_msg(format!("Expected {}, found `{}`", expected, actual))
-}
+type Pair<'a> = ::pest::iterators::Pair<'a, Rule>;
```

Comment:
Sorry, I've not had a chance to read up on pest's design.  What is a `Rule`?  Are we exposing pest in our API via this?

Either way, I think in the first pass, I'd be fine with it and we can experiment and iterate on designs that slowly work to remove pest from the public API.

Reply:
In the first place, this is not a `pub type`, so there is no need to worry about being exposing pest in the API. I just use that type to make function declarations shorter.

`Rule`, if I understand it correctly, is the enum of rules generated by the macro `#[grammar = "grammar.pest"]`, such as `Rule::Expression` or `Rule::Literal` (there is one for each rule defined in `grammar.pest`).

Because `Rule` is only generated in our code, it is bound to the type `pest::iterators::Pair` by a generic type.

Reply:
Maybe I didn't understand what you were asking. Is the `Rule` enum itself on the public API?

Good question, I assumed it wasn't up to this point, but I guess if it is we'll have to isolate it.

Reply:
Macros `#[derive(Parser)]` and `#[grammar = "grammar.pest"]` are now isolated in a module `pest`, so they don't put something unexpected on the public API.

Reply:
So `struct LiquidParser` isn't public, `type Pair` isn't public, and `Rule` (hidden in macro) is not public.

On the other hand, `TagToken` and `TagBlocks` are public and return `Pair` which I assume also returns `Rule`.  That would make them public in every way except being able to refer to their type (which they can refer to `Pair` through pest's API).  Until this is resolved, we should just make them `pub` so people can reference their type.


Weird, Travis isn't reporting any warnings for `private-in-public` which is `warn` by default and out Travis config elevates to `deny`.

Reply:
`TagToken` and `TagBlock`  are public, but they never return `Pair` in their public API, nor I intend them to. I want to keep private every function that would ever return something that exposes `Pest`. The functions that return `Pair` are supposed to be internal helpers only.

Reply:
Sorry, I missed the lack of `pub` on those functions.

Honestly, when implementing the rest of the tags and blocks, I can't see how we could implement them without either these functions being `pub` or effectively implementing every tag and block's parse logic inside of `TagBlock` and `TagTokens`.

Reply:
Again, to be clear, any of this is better than what we have now and we ,an always iterate on it later.

Thinking about that, id go as far as recommending we have pest in the api for now rather than blocking this on trying to find ways to ensure it isn't. Itll give us the benefits of this change sooner and allow us to better see how removing pest impacts the design.

Reply:
To be honest, I think bringing Pest to the API might do more harm than good.

  - Users would need to understand Pest and the underlying structure of the parser in order to understand how to create tags.
  - Bringing `Rule` to the API seems like a bad idea and hard to use because of it being generated by a macro.
  - For `TagTokens`, because many tokens end up being encompassed by `Filterchain`, users would need to manually do the repetitive task of decomposing that token to the one they want. For example, if they desired to expect a `Value`, they would need to access the inner `Pair`s in the `Pair` they've got (the one with `Filterchain` as a rule) and manually test if the input hadn't any unexpected filters, manually returning an error if they had.
  - I suppose it would be harder to implement error handling without requiring the user to constantly manually use functions in their tags that they could easily forget.
  - Removing Pest from the API would imply yet another breaking change.

Given this, I will try to find another way, even if that is simply encapsulating `Pair` itself instead of the iterator over pairs. I suppose that would keep all the flexibility from `Pair` (and maybe even add some utility functions) without exposing the user to something so 'internal' to the parser.

Reply:
> Users would need to understand Pest and the underlying structure of the parser in order to understand how to create tags.

Only tag authors and, as I said, this is a short-term thing.  My priority is to get the benefits of this PR.  We can then experiment on how to refine it from there.

> Removing Pest from the API would imply yet another breaking change.

Yes, we have further breaking changes down the line.  I'm working on some additional ones.  I'd prefer we iterate towards a smaller surface of breaking changes rather than trying to do it all at once.

If nothing else is pushing a release, we could just wait on a release until one or two of these further PRs is in, consolidating these breaking changes.

> Given this, I will try to find another way, even if that is simply encapsulating Pair itself instead of the iterator over pairs. I suppose that would keep all the flexibility from Pair (and maybe even add some utility functions) without exposing the user to something so 'internal' to the parser.

As I said, I'd recommend on getting something in and then having smaller, more targeted PRs to refine how we do this.  Smaller PRs are easier to follow the conversation for, faster to get in, generally less depressing for contributors, etc. 

If you have an idea that is reasonable to execute on within this PR, go for it.  I just recommend we don't try to solve all problems in this PR.

Reply:
I tried to implement all tags with the new API from this PR to see if it was possible to do so without the use of Pairs.

I got stuck, though, as I'm having trouble implementing tags that formerly (in the old parser) depended on `split_block`. I don't think that even after exposing `Pair` in the API I would be able to solve this problem. Do you have any ideas on this?

Reply:
The main one I have is to invert the control flow.  Instead of a block relying on the parser to handle block content and find end of block, we pass the iterator to the block.

In the case of an `if` block, it can iterate until the next tag
- if its `else` or `elsif`, change state and continue.
- if it is an `endif`, then stop processing and give control back to the caller.
- if it is a start tag, delegate the iterator to that block for processing, resuming where it left off when it finishes

This also has the benefit of automatically handling nesting for us.  Instead of tracking nesting depth, we use the call stack.  And `if` in an `if` will consume the iterator until it finds an `endif`, hand control back, and then that `if` will resume from there until it gets to an `endif`.

This does require us to trust the tags/blocks to be written correctly.  Utilities can be provided to help share the termination logic so we don't have to duplicate it, like a `consume_until`.  We might want different styles for different cases rather than a single helper trying to handle all styles of blocks.

As we refactor towards a stable API, these helpers would be moved off into a sub-crate that is free to break compatibility without impacting the use of the top level `liquid` crate.

Reply:
Also, I am fine with you reverting it back to something more akin to the original parser where we `collect` all the elements for a given block which allows the block to do a split.  We could then refactor to my proposal or another idea at a later point.

Reply:
From my quick glance, it seems like it'll be a good starting point.  I'd say, go ahead and polish this and let's get it in.  At some point I'm going to need to pull down your branch to get a better look at it than what I can do within github.